### ISSUE TITLE
support prometheus_client >= 0.4

### ISSUE THREAD

**Issue by richvdh** [matrix-org/synapse#4221]
As a quick fix to #4001 / https://github.com/prometheus/client_python/issues/317, we pinned our prometheus client to < 0.4.0.

This introduces problems:

* Sooner or later we're going to want to pick up some bugfix or feature from more recent prometheus_clients.
* Arch Linux ships with prometheus_client 0.4, so people on Arch have trouble building synapse (https://github.com/matrix-org/synapse/issues/4167). This problem is likely to affect more platforms in the future.
* It's actually possible for people to suffer #4001 in reverse:
  * They start with synapse 0.33.5 or earlier, which doesn't include the pin, so they get prometheus-client 0.4.0, with the `*_total` metrics.
  * Then they upgrade synapse, which includes the pin, so prometheus-client gets downgraded.

I'm still surprised at the prometheus_client devs making the choice to rename these metrics, but I think we're going to have to live with it, so could do with a migration plan.

When we have renamed metrics [in the past](https://github.com/matrix-org/synapse/blob/master/docs/metrics-howto.rst#block-and-response-metrics-renamed-for-0270), we have supported both old and new names for a few releases. That gives people a chance to update their dashboards and alerts. I believe we should do something similar here.

It may turn out that supporting both names is more easily done with prometheus_client 0.3 than it is with 0.4, so we should do this work before we are forced to update to 0.4 in a hurry which would break everyone's metrics.

**Comment by matoro** [comment]
I see NixOS having issues with this, Gentoo also dropped 0.3.1 from the repo about a week ago.

**Comment by nado** [comment]
If we start fresh, is synapse compatible with >0.4.0 ?

**Comment by richvdh** [comment]
> If we start fresh, is synapse compatible with >0.4.0 ?

Not entirely. The tests fail: https://github.com/matrix-org/synapse/issues/4167

**Comment by nado** [comment]
Sad, I think this is my last hardblocker for installing it on my system (gentoo). I don’t really intend on porting <prometheus-client-0.4.0 :/

**Comment by matoro** [comment]
@nado , I have been running with the following as `dev-python/prometheus_client-0.3.1.ebuild`:

```
# Copyright 1999-2018 Gentoo Foundation
# Distributed under the terms of the GNU General Public License v2

EAPI=6

PYTHON_COMPAT=( python{2_7,3_4,3_5,3_6} )
inherit distutils-r1

DESCRIPTION="Python client for the Prometheus monitoring system"
HOMEPAGE="https://pypi.org/project/prometheus_client/"
SRC_URI="https://github.com/prometheus/client_python/archive/v${PV}.tar.gz -> ${P}.tar.gz"

LICENSE="Apache-2.0"
SLOT="0"
KEYWORDS="~amd64"
IUSE="test"

S=${WORKDIR}/client_python-${PV}

RDEPEND="|| ( dev-python/twisted[${PYTHON_USEDEP}] dev-python/twisted-web[${PYTHON_USEDEP}] )"
DEPEND="dev-python/setuptools[${PYTHON_USEDEP}]
        test? ( ${RDEPEND}
                dev-python/pytest[${PYTHON_USEDEP}] )"

python_test() {
        pytest || die
}
```

which seems to work, but then again it shouldn't really matter if you don't use the metrics functionality.

**Comment by nado** [comment]
Yeah I could, but since I intended to make a PR against gentoo tree as a long goal, it seems weird to add back a removed ebuild. It might be okay.

But there’s one thing I don't understand, is it a required or an optional dependency?
https://github.com/matrix-org/synapse/blob/master/synapse/python_dependencies.py#L64 leads me to believe it is required, but your comment makes me thinking otherwise.

**Comment by matoro** [comment]
It is required as specified by the build system.  I just mean that if you don't use those features, it will never be used at runtime.  IMO an ebuild should reflect what upstream says, and only make functionality optional at such time upstream is willing to guarantee that it is written sufficiently modularly.

I have also been running synapse quite well on both x86_64 and armv7a with [this ebuild](https://gist.github.com/matoro/d48d9b2e9365f53f594af3ed97ea2a41) coupled with the gerislay overlay.  This discussion is probably getting a bit off-topic now so if there is another more Gentoo-focused location (issue on an overlay repo?) we should probably use that for further work.

**Comment by matoro** [comment]
FreeBSD is getting hit by this as well:  https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=234828

**Comment by xavierba** [comment]
Fedora is being hit too.
From what I read above, the test failure has been fixed. Is anyone working on the (possibly big) remaining stuff ? Should that be marked as a blocker for 1.0 release ?

**Comment by cyphar** [comment]
[All openSUSE releases now have this problem.](https://build.opensuse.org/package/show/network:messaging:matrix/matrix-synapse) If #4317 fixed the remaining problems with post-`0.4` `prometheus_client` then maybe we should update `python_dependencies.py` so that folks don't hit errors on start with new distributions.

**Comment by hawkowl** [comment]
Fixed in #5636
